### PR TITLE
Add LRScheduler implementation

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
`optim.lr_scheduler.LRScheduler`, `optim.lr_scheduler.ConstantWarmupLRScheduler`, `optim.lr_scheduler.CosineWarmupLRScheduler`, `optim.lr_scheduler.LinearWarmupLRScheduler`, `optim.lr_scheduler.PolyWarmupLRScheduler` classes

Note that for all schedulers implemented, in order to be able to accept `step=0` in the input, the warmup was computed as `x = (train_step_info.step + 1) / (self.total_steps + 1)`, meaning that `+1` was added to both numerator and denominator. This prevents `lr` to be computed as zero in the first iteration (`lr=lr*(0/warmup)`), which would cause all other subsequent lr to be also zero (`lr=0*(x/warmup)`)

### CODE REVIEW

File: orttraining/orttraining/python/training/optim/lr_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
+import math
+
+from . import config
+
+
+class LRScheduler(object):
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/optim/lr_scheduler.py:6]: I am assuming that we expect users to inherit from this base class which is why it is not named with a leading `_` (internal). Is that correct?

File: orttraining/orttraining/python/training/optim/lr_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    r"""Base class for implementing custom learning rate schedulers
+
+    Schedulers can be either statefull or stateless.
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/optim/lr_scheduler.py:9]: nit: spell "stateful".

File: orttraining/orttraining/python/training/optim/lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    Stateless implementations can only rely on information available at
+    :py:class:`.TrainStepInfo` and :py:attr:`.optimizer_config`.
+    Statefull, on the other hand, can store more parameters by overriding
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/optim/lr_scheduler.py:12]: nit: spelling; same as above.

File: orttraining/orttraining/python/training/optim/lr_scheduler.py (hunk 1 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/optim/lr_scheduler.py:6]: correct

File: orttraining/orttraining/python/training/optim/lr_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        r"""Returns a list of learning rate
+
+        Args:
+            train_step_info (:py:class:`.TrainStepInfo`): runtime info for current training step
+
+        Returns:
+            ordered :py:obj:`list` of learning rates.
+                The first entry is the default learning rate and
+                    the remaining, refer to each :py:class:`optim._OptimizerConfig`s parameter group.
+            NOTE: Currently, only the default learning rate is supported,
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/optim/lr_scheduler.py:45]: Checking if we have plans to do anything different in the future?

File: orttraining/orttraining/python/training/optim/lr_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+            "warmup must be a float between (0, 1]"
+        assert total_steps > warmup,\
+            "total_steps must be greater than warmup"
+
+        self.total_steps = total_steps
+        self.warmup = warmup
+
+    def _warmup_constant(self, train_step_info):
+        # Adds 1 to train_step_info.step and self.total_steps to prevent zero'ing lr
+        x = (train_step_info.step + 1) / (self.total_steps + 1)
```
Comment (spandantiwari) [inline orttraining/orttraining/python/training/optim/lr_scheduler.py:111]: How does this warmup strategy (especially the +1 in num and den) compare with PyTorch?

File: orttraining/orttraining/python/training/optim/lr_scheduler.py (hunk 4 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/optim/lr_scheduler.py:45]: After talking to Sherlock and Liqun, we agreed this feature is possible through some changes on ORT backend and it might be requested by users. Therefore, the new frontend does support multiple lr to be specified, but we are blocking it until backend provides support

File: orttraining/orttraining/python/training/optim/lr_scheduler.py (hunk 5 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/optim/lr_scheduler.py:111]: Pytorch has a different approach and don't use `step` we do, probably because their scheduler don't use the warmup concept.

In their approach, they rely on `epoch` instead. When `epoch > 0`, compute is done using `epoch` as part of the input, otherwise, the initial lr is returned without any compute (reference https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L310)

File: orttraining/orttraining/python/training/optim/lr_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+
+class _LRScheduler(object):
+    r"""Base class for implementing custom learning rate schedulers
+
+    Schedulers can be either stateful or stateless.
+    Stateless implementations can only rely on information available at
+    :py:class:`.TrainStepInfo` and :py:attr:`.optimizer_config`.
+    Stateful, on the other hand, can store more parameters by overriding
+    the constructor.
```
Comment (BowenBao) [inline orttraining/orttraining/python/training/optim/lr_scheduler.py:13]: Could we claim that for a given initialized stateless scheduler, for a particular instance of `TrainStepInfo` object, the `get_lr` method will always output the same result. 

stateful, on the other hand, might return different results each time even called with the same `TrainStepInfo` object.

File: orttraining/orttraining/python/training/optim/lr_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        r"""Private method called to update :py:class:`optim._OptimizerConfig` learning rate
+
+        NOTE: This class should never be called by the user.
+        """
+        new_lr = self.get_lr(train_step_info)
+        self._last_lr = new_lr
+        self.optimizer_config.lr = new_lr[0]
+
+
+class ConstantWarmupLRScheduler(_LRScheduler):
```
Comment (BowenBao) [inline orttraining/orttraining/python/training/optim/lr_scheduler.py:64]: For the following LRSchedulers, is there a way to tell if they are stateful or stateless?

File: orttraining/orttraining/python/training/optim/lr_scheduler.py (hunk 5 above)
Comment (BowenBao) [reply orttraining/orttraining/python/training/optim/lr_scheduler.py:111]: Are we able to create lr schedulers that match PyTorch behavior?

Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/optim/lr_scheduler.py:111]: Yes, it should as both `epoch` and `step` are available at `LRSCheduler.update(self, train_step_info)` and other scheduler specific parameters can be passed in to the LRScheduler constructor.

File: orttraining/orttraining/python/training/optim/lr_scheduler.py (hunk 6 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/optim/lr_scheduler.py:13]: Generally speaking, yes, as a consequence of stateless vs stateful definition. But It is possible that even stateless implementation can use random generatorts initialized by "current timestamp" and other stuff that are not stored anywhere, but can be used as an input for the scheduling algorithm

Removing all this randomness aforementioned, stateless implementation can output the same result given a static `TrainStepInfo` *and* `_OptimizerConfig`

File: orttraining/orttraining/python/training/optim/lr_scheduler.py (hunk 7 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/optim/lr_scheduler.py:64]: The class description and/or implementation are the only official way
Looking the LRSCheduler constructor we can make an educated guess. A stateless class doesn't need to receive any input in the constructor, as it shouldnt store state from external sources. However, that doesnt mean the LRScheduler is not initializing state internally without external information.
